### PR TITLE
Enable linking of encoders to switch within layout macros

### PR DESCRIPTION
**Pull request by zvecr**
## Description

Will allow XAP client (or configurator) to know a given key should be rendered as circle, and index for addressing into encoder map.

Adds crude `qmk info` rendering where drawing circles is super fun.... but its at least something.
image

Also adds validation to ensure the values are somewhat correct.

* Additionally add something `"type": "switch|encoder|encoder_switch"`  instead of just inferring from index key?
  * Could be done on api gen?
* Future, extend this to rotation only encoders? 
  * Could strip keys with `encoder:x` when generating layout macros?
    * Would break current assumption that all must have matrix
      * Would need strategy on how to avoid user config errors?
<details>
  <summary> Alternatives? </summary>

### Add layout as `encoders.layout[]`?
```json
{
    "encoder": {
        "rotary": [
            {"pin_a": "B5", "pin_b": "B4"},
            {"pin_a": "B3", "pin_b": "B2"}
        ]
        "layout": [
            { "x": 8, "y": 60, "matrix": [5, 0] },
            { "x": 11, "y": 40 },
        ]
    }
}
```

* Does solve non-click encoders
* Would cause a lot of complexity with merging into regular layouts
* Would need to be an array like `layouts`?
  * Optional encoders? Same encoder could be physically laid out to multiple locations?
  * current proposal could handle this with just additional layout macros

</details>

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by zvecr** [comment]
Quick hack about of configurator to use new property to render key as encoder
image

File: lib/python/qmk/keyboard.py [hunk 1]
```diff
@@ -30,6 +30,28 @@
...
+ENC_DRAWING_CHARACTERS = {
+    "unicode": {
+        "tl": "🮣",
+        "tr": "🮢",
+        "bl": "🮡",
+        "br": "🮠",
+        "vl": "│",
+        "vr": "│",
+        "v": "│",
+        "h": "─",
```
Comment (tzarc) [inline lib/python/qmk/keyboard.py:42]: how about these?
```suggestion
        "tl": "╭",
        "tr": "╮",
        "bl": "╰",
        "br": "╯",
        "vl": "│",
        "vr": "│",
        "v": "│",
        "h": "─",
```

Comment (zvecr) [reply lib/python/qmk/keyboard.py:42]: Already tried, visually they are not different enough to the regular key.

**Review by tzarc — APPROVED** [review]

File: lib/python/qmk/keyboard.py (hunk 1 above)
Comment (zvecr) [reply lib/python/qmk/keyboard.py]: Windows being windows, ive gone with the rounded corners and some arrows. Someone else can improve this later if they wish.

**Comment by fauxpark** [comment]
What about the case where an encoder does not have a key associated with it?

Also, I'm planning to completely rework the CLI rendering (to look more like the matrix diagrams that have been added recently), which would not work with the way this is implemented.

**Comment by zvecr** [comment]
> What about the case where an encoder does not have a key associated with it?

Mentioned in PR description,
> Future, extend this to rotation only encoders?

Adding stuff right now would break configurator, so probably fine for now as that case isnt that widespread.

> Also, I'm planning to completely rework the CLI rendering (to look more like the matrix diagrams that have been added recently), which would not work with the way this is implemented.

All i've done is copy the existing implementation pattern, and have zero care if it is later thrown away. This shouldnt be held up for something that *might* happen in the future, given this has been a prerequisite of XAP for a long time.

**Comment by fauxpark** [comment]
I didn't say it should be held up.

**Review by tzarc — APPROVED** [review]

**Pull request merged**